packages/bulkapi/src/database/DbChanges.ts

```diff
@@ -0,0 +1,314 @@
+                    }
+                    this.updatesOnNodeTable.add(update.id, update)
+                    break
+                }
+                case "PropertyValueChanged": {
+                    const update: DbPropertyUpdate = {
+                        node_id: (change as PropertyValueChanged).nodeId,
+                        property: (change as PropertyValueChanged).property,
+                        column: "value",
+                        newValue: (change as PropertyValueChanged).newValue,
+                        missing: (change as PropertyValueChanged).missing
+                    }
+                    this.updatesPropertyTable.add({ nodeId: update.node_id, property: update.property }, update)
+                    break
+                }
+                case "ChildRemoved":
+                case "ChildAdded":
+                case "ChildOrderChanged": {
+                    const update: DbContainmentUpdate = {
+                        node_id: (change as ContainmentChange).parentNode.id,
+                        containment: (change as ContainmentChange).containment,
+                        column: "children",
+                        children: (change as ContainmentChange).afterContainment?.children ?? [],
+                        missing: (change as ContainmentChange).missing 
+                    }
+                    this.updatesContainmentTable.add({ nodeId: update.node_id, containment: update.containment }, update)
+                    break
+                }
+                case "TargetAdded":
+                case "TargetRemoved":
+                case "TargetOrderChanged": {
+                    const update: DbReferenceUpdate = {
+                        node_id: (change as ReferenceChange).node.id,
+                        reference: (change as ReferenceChange).beforeReference.reference,
+                        column: "targets",
+                        targets: (change as ReferenceChange).afterReference?.targets ?? [],
+                        missing: (change as ReferenceChange).missing
+                    }
+                    this.updatesReferenceTable.add({ nodeId: update.node_id, reference: update.reference }, update)
+                    break
+                }
+            }
+        })
+    }
+
+    /**
+     * Create a Postgres query for all the changes added to this DbCommand.
+     */
+    createPostgresQuery(): string {
+        let result = `-- Update generated by DbCommands\n`
+        this.updatesOnNodeTable.values().forEach((values: DbNodeUpdate[]) => {
+            const newValue = {}
+            values.forEach(v => (newValue[v.column] = v.newValue))
+            result += `-- update nodes
+                        UPDATE ${NODES_TABLE}
+                            SET ${this.context.pgp.helpers.sets(newValue, Object.keys(newValue))}
+                        WHERE
+                            id = '${values[0].id}';
```

Comment:
Are we sure these IDs conform to the specification? In particular: is this SQL-injectable? Is there a way to compose a text for this query that is fundamentally un-injectable, e.g. using a `VALUES` clause?

Reply:
The chunk is validated before this code is ever run, so it does conform to the specification. Assuming the validation is working correctly of course.

Reply:
Is VALUES less a problem with SQL injection?
By the way, the "helper.sets" creates the actual text for the query and is part of the pg-promise library.  It does checks on the values provided and ensures everything is escaped if needed.

Reply:
`VALUES` is a different syntax so it's less likely to offer an attack surface for “Bobby Tables” in the same way that a direct injection would. But attackers could try with both syntaxes of course... Maybe it's just that I personally prefer the `VALUES` syntax...